### ISSUE TITLE
breaking changes in 1.12

### ISSUE THREAD

**Issue by kakulukia** [wagtail/wagtail#3995]
Hi!
 i was just updating to the latest version only to find out that the rending of StreamField block kinda changed in a way, that its broken for me now. Reverting back to my old version, i was able to find out that 1.11.1 displays my blocks as they should be and 1.12 introduces breaking changes resulting in output like this:

    StructValue([('column_width', '4'), ('elements', [<wagtail.wagtailcore.blocks.stream_block.StreamValue.StreamChild object at 0x10d83c940>, <wagtail.wagtailcore.blocks.stream_block.StreamValue.StreamChild object at 0x10d83cac8>])])

Reading through the release notes of 1.12 i cant find anything related to breaking changes in StreamField rendering. Going back to the old version fixes the rendering problem again, but id like to update and and fix the issue. What happened/changed here?

**Comment by gasman** [comment]
Hi @kakulukia,
This is most likely due to #3738. This problem should only arise if you're doing something with the HTML representation of a StructBlock _outside_ of Django's template engine - for example, you're using a custom template tag to perform text replacement on it.

If this is occurring in something that seems to be perfectly ordinary Wagtail code, please provide a _minimal_ example reproducing the issue, and we'll investigate further.

**Comment by kakulukia** [comment]
nope @gasman , i dont do anything like that. I have this row block:

    RowBlock = blocks.ListBlock(ColumnBlock(), template='cms/blocks/row_stream.pug', icon='fa-table')

rendering its contents that way:

    .row
      for element in self
        | {{ element }}

**Comment by kakulukia** [comment]
The ColumnBlock looks this way:

    class ColumnBlock(blocks.StructBlock):
    column_width = blocks.ChoiceBlock(choices=[ ... ])

    class Meta:
        template = 'cms/blocks/column_block.pug'

But whatever i change in its template does not affect the rendering problem, so i assume its the above RowBlock causing problems.

**Comment by gasman** [comment]
The `.pug` template engine you're using has never been supported by Wagtail. When I asked if you were doing something non-standard, this would have been an important detail to mention... ;-)

Sorry, you'll have to troubleshoot this yourself.

**Comment by kakulukia** [comment]
Well @gasman, i just changed the template from pug to default html


    <div class="row">
      {% for element in self %}
        {{ element }}
      {% endfor %}
    </div>

But that does not change whats happening - and as i said the breaking change was introduced with 1.12 rather than another module update. And it can be reversed lowering the wagtail version again to 1.11.1.

**Comment by kbrownlees** [comment]
I just had the same issue when trying to upgrade a wagtail project to 1.13.1 (using django templates). 'Fixed' it by making the template call {{ element.render_as_block }}.

In my case it was iterating a number of columns which were being added to the context manually:

```
{% for column in columns %}
<div class="col-xs-6 col-sm-4 col-md-2">
    <div class="s-f-m">
        {{ column.render_as_block }}
    </div>
</div>
{% endfor %}
```

And get_context (simplified):
```
class FooterBlock(StructBlock):
    column_two = FooterColumnBlock(label='Second column content')
    column_three = FooterColumnBlock(label='Third column content')

    class Meta:
        template = 'common/blocks/footer.html'
        icon = 'fa-th'
        label = 'Footer menu'

    def get_context(self, value, parent_context=None):
        context = super(FooterBlock, self).get_context(value, parent_context=parent_context)
        context['columns'] = [
            value['column_two'], value['column_three']
        ]
        return context
```

Probably a non standard way of doing things, but might be worth a mention in the release notes?

**Comment by kakulukia** [comment]
oh, reminding me of that ticket .. i was able to fix my problem while exchanging {{ element }} with {% include_block element %}
i still dont know why this is happening, tho
@gasman looking at the workarounds, do you have a clue now what might cause that?

**Comment by gasman** [comment]
It looks like #3738 is the direct cause of this - we had to stop using the template as the `str()` representation of the struct value, because that was causing infinite loops on some logging/debugging tools. Originally I thought this wouldn't affect template rendering, because we define a `__html__()` method which _does_ render the template - but it seems that this isn't being picked up.

Unfortunately I'm not aware of any way we can make `{{ element }}` work without defining a `__str__()` method (and thus reintroducing the issue fixed in #3738). If someone's willing to dig into Django's template logic to find another way of making it work, I'd be very interested to hear the results.

Failing that, I think we have to leave this unfixed - the recommended workaround is to use `{% include_block element %}` (and this is the preferred method of including blocks that use templates, because it allows access to context variables in the calling page).
